### ISSUE TITLE
NumberField to preserve decimals

### ISSUE THREAD

**Issue by StefanCov** [vaadin/vaadin-text-field-flow#219]
When settings the number field to step by 1.00d it will lose its decimals. For example, 1.98 will jump to 2 but I would like it to jump to 2.98. If there was a way to know if the change event was a button press then I could place the last digits of the oldValue onto the new value. If some kind of method was introduced to preserve the decimals and jump by the exact number stated without any kind of rounding would be very handy in multiple circumstances.

**Comment by tomivirkki** [comment]
Hi. It works the same ways as native `<input type="number">` because that's what it uses internally. One way to make the number field behave as you suggested would be to adjust the step base value on every value change by modifying the `min` property of the field.

For example if the current value is 0.5 and step is 1, when you hit up, the value goes to 1 as you described.
But if you change the min value also to be for example -10.5, that will be the new step base for the field and hitting up from 0.5 goes to 1.5.

**Comment by StefanCov** [comment]
Sorry I'm a bit confused by what you mean. So you're saying to change the min value on the ValueChangeEvent? Could you provide a small basically example?

**Comment by tomivirkki** [comment]
Here's a quick demo of what I meant (in JS): https://jsbin.com/gukuvifiri/1/edit?html,output

**Comment by StefanCov** [comment]
`NumberField selector = new NumberField();`
`selector.setValue(2.98);`
`selector.setHasControls(true);`
`selector.setStep(1);`
`selector.addValueChangeListener(e -> {`
       ` selector.setMin((selector.getValue()-selector.getStep()*10));`
`});`

I tried this but had no luck in getting the result you have in your js example...

**Comment by tomivirkki** [comment]
This works for me, changes:
- Set the initial value after setting the valuechangelistener
- Use eager value change mode (so you don't need to blur the field for the listener to trigger)
```
NumberField selector = new NumberField();
selector.setHasControls(true);
selector.setStep(1);
selector.setValueChangeMode(ValueChangeMode.EAGER);
selector.addValueChangeListener(e -> {
    selector.setMin((selector.getValue()-selector.getStep()*10));
});
selector.setValue(2.98);
```

**Comment by StefanCov** [comment]
This works! Thanks!

**Comment by StefanCov** [comment]
I know this issue has been closed but I have come across another problem with having the setValue after the listener. When using the numberField in a Vaadin Grid it causes issues if the valueChangeListener is used to refresh the dataProvider item. This causes a unbreakable loop, so I stopped this by only calling a dataProvider refresh when it is by client input. This causes another issue as each number typed causes the row to refresh, which means i'm only able to type 1 number at a time before the row refreshes and it takes me off number field.
